core/src/main/java/org/pentaho/di/compatibility/Value.java

```diff
@@ -0,0 +1,3960 @@
+
+  public Value num2str() throws KettleValueException {
+    return num2str( null, null, null, null );
+  }
+
+  public Value num2str( String format ) throws KettleValueException {
+    return num2str( format, null, null, null );
+  }
+
+  public Value num2str( String format, String decimalSymbol ) throws KettleValueException {
+    return num2str( format, decimalSymbol, null, null );
+  }
+
+  public Value num2str( String format, String decimalSymbol, String groupingSymbol ) throws KettleValueException {
+    return num2str( format, decimalSymbol, groupingSymbol, null );
+  }
+
+  public Value num2str( String format, String decimalSymbol, String groupingSymbol, String currencySymbol ) throws KettleValueException {
+    if ( isNull() ) {
+      setType( VALUE_TYPE_STRING );
+    } else {
+      // Number to String conversion...
+      if ( getType() == VALUE_TYPE_NUMBER || getType() == VALUE_TYPE_INTEGER ) {
+        NumberFormat nf = NumberFormat.getInstance();
+        DecimalFormat df = (DecimalFormat) nf;
+        DecimalFormatSymbols dfs = new DecimalFormatSymbols();
+
+        if ( currencySymbol != null && currencySymbol.length() > 0 ) {
+          dfs.setCurrencySymbol( currencySymbol );
+        }
+        if ( groupingSymbol != null && groupingSymbol.length() > 0 ) {
+          dfs.setGroupingSeparator( groupingSymbol.charAt( 0 ) );
+        }
+        if ( decimalSymbol != null && decimalSymbol.length() > 0 ) {
+          dfs.setDecimalSeparator( decimalSymbol.charAt( 0 ) );
+        }
+        df.setDecimalFormatSymbols( dfs ); // in case of 4, 3 or 2
+        if ( format != null && format.length() > 0 ) {
+          df.applyPattern( format );
+        }
+        try {
+          setValue( nf.format( getNumber() ) );
+        } catch ( Exception e ) {
+          setType( VALUE_TYPE_STRING );
+          setNull();
+          throw new KettleValueException( "Couldn't convert Number to String " + e.toString() );
+        }
+      } else {
+        throw new KettleValueException( "Function NUM2STR only works on Numbers and Integers" );
+      }
+    }
+    return this;
+  }
+
+  // date: TO_CHAR( <date> , 'yyyy/mm/dd HH:mm:ss'
+  public Value dat2str() throws KettleValueException {
+    return dat2str( null, null );
+  }
+
+  public Value dat2str( String arg0 ) throws KettleValueException {
+    return dat2str( arg0, null );
+  }
+
+  public Value dat2str( String arg0, String arg1 ) throws KettleValueException {
+    if ( isNull() ) {
+      setType( VALUE_TYPE_STRING );
+    } else {
+      if ( getType() == VALUE_TYPE_DATE ) {
+        SimpleDateFormat df = new SimpleDateFormat();
+
+        DateFormatSymbols dfs = new DateFormatSymbols();
+        if ( arg1 != null ) {
+          dfs.setLocalPatternChars( arg1 );
+        }
+        if ( arg0 != null ) {
+          df.applyPattern( arg0 );
+        }
+        try {
+          setValue( df.format( getDate() ) );
+        } catch ( Exception e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)